File: 437.路径总和-iii.py
#
# @lc app=leetcode.cn id=437 lang=python3
#
# [437] 路径总和 III
#
# https://leetcode-cn.com/problems/path-sum-iii/description/
#
# algorithms
# Medium (56.43%)
# Likes:    683
# Dislikes: 0
# Total Accepted:    60.1K
# Total Submissions: 106.5K
# Testcase Example:  '[10,5,-3,3,2,null,11,3,-2,null,1]\n8'
#
# 给定一个二叉树，它的每个结点都存放着一个整数值。
# 
# 找出路径和等于给定数值的路径总数。
# 
# 路径不需要从根节点开始，也不需要在叶子节点结束，但是路径方向必须是向下的（只能从父节点到子节点）。
# 
# 二叉树不超过1000个节点，且节点数值范围是 [-1000000,1000000] 的整数。
# 
# 示例：
# 
# root = [10,5,-3,3,2,null,11,3,-2,null,1], sum = 8
# 
# ⁠     10
# ⁠    /  \
# ⁠   5   -3
# ⁠  / \    \
# ⁠ 3   2   11
# ⁠/ \   \
# 3  -2   1
# 
# 返回 3。和等于 8 的路径有:
# 
# 1.  5 -> 3
# 2.  5 -> 2 -> 1
# 3.  -3 -> 11
# 
# 
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        def helper(node: TreeNode, pre : int, depth: int) :
            if not node: return
            if node.val == pre: self.ans += 1
            # choose this node
            helper(node.left, pre - node.val, 0)
            helper(node.right, pre - node.val, 0)
            # not choose this node
            # Represents the depth of the first node that is not taken
            # In fact, it is to judge whether the root node is not taken
            if depth:
                helper(node.left, sum, 1)
                helper(node.right, sum, 1)
        self.ans = 0
        helper(root, sum, 1)
        return self.ans
            
# @lc code=end

File: 437.路径总和-iii.cs
/*
 * @lc app=leetcode.cn id=437 lang=csharp
 *
 * [437] 路径总和 III
 *
 * https://leetcode-cn.com/problems/path-sum-iii/description/
 *
 * algorithms
 * Medium (56.43%)
 * Likes:    683
 * Dislikes: 0
 * Total Accepted:    60.1K
 * Total Submissions: 106.5K
 * Testcase Example:  '[10,5,-3,3,2,null,11,3,-2,null,1]\n8'
 *
 * 给定一个二叉树，它的每个结点都存放着一个整数值。
 * 
 * 找出路径和等于给定数值的路径总数。
 * 
 * 路径不需要从根节点开始，也不需要在叶子节点结束，但是路径方向必须是向下的（只能从父节点到子节点）。
 * 
 * 二叉树不超过1000个节点，且节点数值范围是 [-1000000,1000000] 的整数。
 * 
 * 示例：
 * 
 * root = [10,5,-3,3,2,null,11,3,-2,null,1], sum = 8
 * 
 * ⁠     10
 * ⁠    /  \
 * ⁠   5   -3
 * ⁠  / \    \
 * ⁠ 3   2   11
 * ⁠/ \   \
 * 3  -2   1
 * 
 * 返回 3。和等于 8 的路径有:
 * 
 * 1.  5 -> 3
 * 2.  5 -> 2 -> 1
 * 3.  -3 -> 11
 * 
 * 
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    int ans = 0;
    int sum;
    public int PathSum(TreeNode root, int sum) {
        this.sum = sum;
        Helper(root, sum, true);
        return ans;
    }
    private void Helper(TreeNode node, int pre, bool isFirst) {
        if(node == null) return;
        if(node.val == pre) ans++;
        Helper(node.left, pre - node.val, false);
        Helper(node.right, pre - node.val, false);
        if(isFirst) {
            Helper(node.left, sum, true);
            Helper(node.right, sum, true);
        }
    }
}
// @lc code=end

